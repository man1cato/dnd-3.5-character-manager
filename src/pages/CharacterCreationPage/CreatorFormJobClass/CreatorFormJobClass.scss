@import '../CharacterCreationPage.scss';


.CreatorFormJobClass__group {
   @extend .CreatorForm__group;
   > :first-child {
      width: 35%;
      padding-right: $s-size;
   }
   > :last-child {
      width: 65%;
   }
   button {
      max-width: 80px;
   }
}

.CreatorFormJobClass__group--top {
   @extend .CreatorFormJobClass__group;
   align-items: flex-start;
} 