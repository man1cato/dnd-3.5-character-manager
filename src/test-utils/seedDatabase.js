import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}