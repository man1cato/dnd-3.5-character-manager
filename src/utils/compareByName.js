export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}