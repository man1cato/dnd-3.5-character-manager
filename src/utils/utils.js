import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(_.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])


export const rollDice = (dieFaces = 6, countDice = 1) => {
	let total = 0
	for (let i = 0; i < countDice; i++) {
		total += Math.floor((Math.random() * dieFaces) + 1)
	}
	return total
}

export const calcAbilityMod = (abilityScore) => Math.floor(abilityScore/2 - 5)

export const calSizeMod = (size) => ({
	Huge: -2,
	Large: -1,
	Medium: 0,
	Small: 1,
	TIny: 2
}[size])

export const calcSkillPoints = (jobClassName, intMod) => {
	if (_.includes(['Cleric', 'Fighter', 'Paladin', 'Sorcerer', 'Wizard'], jobClassName)) {
		return 2 + intMod
	}
	if (_.includes(['Barbarian', 'Druid', 'Monk'], jobClassName)) {
		return 4 + intMod
	}
	if (_.includes(['Bard', 'Ranger'], jobClassName)) {
		return 6 + intMod
	}
	if (jobClassName === 'Rogue') {
		return 8 + intMod
	}
}