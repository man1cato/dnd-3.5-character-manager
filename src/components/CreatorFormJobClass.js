import React, { useState, useEffect } from 'react'
import _ from 'lodash'
import { Field, ErrorMessage } from 'formik'
import { schools, alignments } from '../utils/staticData'
import { apiObjectToArray } from '../utils/utils'


const validateProhibitedSchools = (selectedSchool) => (value) => {
	let error;
	if (selectedSchool === 'Divination') {
		if (value.length !== 1) {
			error = 'You must select one school whose magic use will be prohibited';
		}
	} else {
		if (selectedSchool !== 'Universal' && value.length !== 2) {
			error = 'You must select two schools whose magic use will be prohibited';
		}
	}
	return error;
}

const CreatorFormJobClass = ({
	values, 
	races,
	jobClasses, 
	handleChange, 
	handleMultiSelect, 
	setFieldValue, 
	setFieldError,
	validateForm
}) => {
	const [selectedJobClass, setSelectedJobClass] = useState(jobClasses[values.jobClass])

	useEffect(() => {
		if (selectedJobClass.name === 'Paladin') {
			setFieldValue('alignment', 'Lawful Good')
		} else if (selectedJobClass.name === 'Wizard') {
			setFieldValue('school', 'Universal')
		} else {
			setFieldValue('school', null)
			setFieldValue('prohibitedSchools', [])
		}
	}, [selectedJobClass])

	return (
	<>
		<div className="form-group--35">
			<h4>Job Class:</h4>
			<Field
				className="select"
				name="jobClass"
				component="select"
				onChange={(e) => {
					setSelectedJobClass(jobClasses[e.target.value])
					handleChange(e)
				}}
			>
				{apiObjectToArray(jobClasses).map((jobClass, i) => (
					<option
						key={`jobClass${i}`}
						value={jobClass.id}
					>
						{jobClass.name}
					</option>
				))}
			</Field>
		</div>

		<div className="form-group--35">
			<h4>Alignment:</h4>
			{selectedJobClass.name === "Paladin" ?
				<div>Lawful Good</div>
				:
				<Field 
					className="select" 
					name="alignment" 
					component="select" 
				>
					{alignments.map((alignment, i) => (
						<option 
							value={alignment} 
							key={`alignment${i}`}
						>
							{alignment}
						</option>
					))}
				</Field>
			}
		</div>		

		<div className="form-group--35 align-top">
			<h4>Hit Die:</h4>
			<div>{selectedJobClass.hitDie}</div>
		</div>

		<div className="form-group--35 align-top">
			<h4>Proficiencies:</h4>
			<div>{selectedJobClass.proficiencies.join(", ")}</div>
		</div>

		<div className="form-group--35 align-top">
			<h4>Bonus Languages:</h4>
			<Field 
				className="multi-select" 
				name="bonusLanguages" 
				component="select" 
				multiple 
				onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
			>
				{_.orderBy(races[values.race].bonusLanguages).map((language) => (
					<option 
						value={language} 
						key={language}
					>
						{language}
					</option>
				))}
			</Field>                                    
		</div>

		<div className="form-group--35">
			<h4>Deity:</h4>
			<Field 
				className="text-input" 
				name="deity" 
				placeholder="(Optional) Enter the name of a Deity" 
			/>                                  
		</div>
		
		{selectedJobClass.name === 'Wizard' && (
			<div className="form-group--35">
				<h4>School:</h4>
				<Field 
					className="select" 
					name="school" 
					component="select" 
					onChange={(e) => {
						if (e.target.value === 'Universal') { 
							setFieldError('prohibitedSchools', undefined)	
							setFieldValue('prohibitedSchools', [])
						}						
						handleChange(e)
					}}
				>
					{schools.map((school, i) => (
						<option 
							value={school} 
							key={`school${i}`}
						>
							{school}
						</option>
					))}
				</Field>
			</div>
		)}

		{selectedJobClass.name === 'Wizard' && values.school !== 'Universal' && (
			<>
				<div className="form-group--35 align-top">
					<h4>Prohibited School(s):</h4>
					<Field 
						className="multi-select"
						name="prohibitedSchools" 
						component="select" 
						multiple 
						validate={validateProhibitedSchools(values.school)}
						onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
					>
						{_.difference(schools, [values.school, 'Universal', 'Divination']).map((school, i) => (
							<option 
								value={school} 
								key={`prohibited${i}`}
							>
								{school}
							</option>
						))}
					</Field>
				</div>
				<ErrorMessage className="form-group--error" name="prohibitedSchools" component="div"/>
			</>
		)}
	</>
)
}

export default CreatorFormJobClass
