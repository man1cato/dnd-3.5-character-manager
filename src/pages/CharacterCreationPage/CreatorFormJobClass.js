import React, { useEffect } from 'react'
import _ from 'lodash'
import { Field, ErrorMessage } from 'formik'

import { schools, alignments } from '../../utils/staticData'
import { apiObjectToArray, calcStartingGold } from '../../utils/utils'


const validateProhibitedSchools = (selectedSchool) => (value) => {
	let error
	if (selectedSchool === 'Divination') {
		if (value.length < 1) {
			error = 'You must select one school whose magic use will be prohibited'
		}
		if (value.length > 1) {
			error = 'Select only one school whose magic use will be prohibited'
		}
	} else {
		if (selectedSchool !== 'Universal' && value.length < 2) {
			error = 'You must select two schools whose magic use will be prohibited'
		}
		if (selectedSchool !== 'Universal' && value.length > 2) {
			error = 'Select only two schools whose magic use will be prohibited'
		}
	}
	return error
}

const CreatorFormJobClass = ({
	values, 
	selectedRace,
	selectedJobClass, 
	setSelectedJobClass,
	jobClasses, 
	handleChange, 
	handleMultiSelect, 
	setFieldValue, 
	setFieldError
}) => {
	const { school } = values

	useEffect(() => {
		if (selectedJobClass.name === 'Paladin') {
			setFieldValue('alignment', 'Lawful Good')
		} else if (selectedJobClass.name === 'Wizard') {
			setFieldValue('school', 'Universal')
		} else {
			setFieldValue('school', null)
			setFieldValue('prohibitedSchools', [])
		}
	}, [selectedJobClass])

	return (
		<div className="container--body">			
			<div className="form-group--35">
				<h4>Job Class:</h4>
				<Field
					className="select"
					name="jobClass"
					component="select"
					onChange={(e) => {
						setSelectedJobClass(jobClasses[e.target.value])
						setFieldValue('startingGold', 0)
						handleChange(e)
					}}
				>
					{apiObjectToArray(jobClasses).map((jobClass, i) => (
						<option
							key={`jobClass${i}`}
							value={jobClass.id}
						>
							{jobClass.name}
						</option>
					))}
				</Field>
			</div>

			<div className="form-group--35">
				<h4>Alignment:</h4>
				{selectedJobClass.name === "Paladin" ?
					<div>Lawful Good</div>
					:
					<Field 
						className="select" 
						name="alignment" 
						component="select" 
					>
						{alignments.map((alignment, i) => (
							<option 
								value={alignment} 
								key={`alignment${i}`}
							>
								{alignment}
							</option>
						))}
					</Field>
				}
			</div>		

			<div className="form-group--35 align-top">
				<h4>Hit Die:</h4>
				<div>{selectedJobClass.hitDie}</div>
			</div>

			<div className="form-group--35">
				<h4>Starting Gold:</h4>
				<div className="input-group">
					<input
						className="number-input"
						type="number"
						value={values.startingGold}
						onChange={(e) => setFieldValue('startingGold', e.target.value)}
					/>
					<div> gp</div>
					<button
						className="button"
						type="button"
						onClick={() => setFieldValue('startingGold', calcStartingGold(selectedJobClass.name))}
					>
						Roll
					</button>
				</div>
			</div>

			<div className="form-group--35 align-top">
				<h4>Proficiencies:</h4>
				<div>{selectedJobClass.proficiencies.join(", ")}</div>
			</div>

			<div className="form-group--35 align-top">
				<h4>Bonus Languages:</h4>
				<Field 
					className="multi-select" 
					name="bonusLanguages" 
					component="select" 
					multiple 
					onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
				>
					{_.orderBy(selectedRace.bonusLanguages).map((language) => (
						<option 
							value={language} 
							key={language}
						>
							{language}
						</option>
					))}
				</Field>                                    
			</div>

			<div className="form-group--35">
				<h4>Deity:</h4>
				<Field 
					className="text-input" 
					name="deity" 
					placeholder="(Optional) Enter the name of a Deity" 
				/>                                  
			</div>
		
			{selectedJobClass.name === 'Wizard' && (
				<div className="form-group--35">
					<h4>School:</h4>
					<Field 
						className="select" 
						name="school" 
						component="select" 
						onChange={(e) => {
							if (e.target.value === 'Universal') { 
								setFieldError('prohibitedSchools', undefined)	
								setFieldValue('prohibitedSchools', [])
							}						
							handleChange(e)
						}}
					>
						{schools.map((school, i) => (
							<option 
								value={school} 
								key={`school${i}`}
							>
								{school}
							</option>
						))}
					</Field>
				</div>
			)}

			{selectedJobClass.name === 'Wizard' && school !== 'Universal' && (
				<>
					<div className="form-group--35 align-top">
						<h4>Prohibited School(s):</h4>
						<Field 
							className="multi-select"
							name="prohibitedSchools" 
							component="select" 
							multiple 
							validate={validateProhibitedSchools(school)}
							onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
						>
							{_.difference(schools, [school, 'Universal', 'Divination']).map((school, i) => (
								<option 
									value={school} 
									key={`prohibited${i}`}
								>
									{school}
								</option>
							))}
						</Field>
					</div>
					<ErrorMessage className="form-group--error" name="prohibitedSchools" component="div"/>
				</>
			)}
		</div>
	)
}

export default CreatorFormJobClass
