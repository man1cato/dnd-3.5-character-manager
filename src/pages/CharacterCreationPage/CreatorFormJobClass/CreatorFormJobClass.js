import React, { useEffect } from 'react'
import _ from 'lodash'
import { Field, ErrorMessage } from 'formik'

import { schools, alignments } from '../../../utils/staticData'
import { apiObjectToArray, calcStartingGold } from '../../../utils/utils'
import './CreatorFormJobClass.scss'


const validateProhibitedSchools = selectedSchool => value => {
	let error
	if (selectedSchool === 'Divination') {
		if (value.length < 1) {
			error = 'You must select one school whose magic use will be prohibited'
		}
		if (value.length > 1) {
			error = 'Select only one school whose magic use will be prohibited'
		}
	} else {
		if (selectedSchool !== 'Universal' && value.length < 2) {
			error = 'You must select two schools whose magic use will be prohibited'
		}
		if (selectedSchool !== 'Universal' && value.length > 2) {
			error = 'Select only two schools whose magic use will be prohibited'
		}
	}
	return error
}

const CreatorFormJobClass = ({
	values, 
	jobClasses, 
	handleChange, 
	handleMultiSelect, 
	setFieldValue, 
	setFieldError
}) => {
	const { race, jobClass, school } = values

	useEffect(() => {
		if (jobClass.name === 'Paladin') {
			setFieldValue('alignment', 'Lawful Good')
		} else if (jobClass.name === 'Wizard') {
			setFieldValue('school', 'Universal')
		} else {
			setFieldValue('school', null)
			setFieldValue('prohibitedSchools', [])
		}
	}, [jobClass])

	return (
		<div className="container--body">			
			<div className="CreatorFormJobClass__group">
				<h4>Job Class:</h4>
				<select
					className="select"
					defaultValue={_.findKey(jobClasses, jobClass)}
					onChange={(e) => {
						setFieldValue('jobClass', jobClasses[e.target.value])
						setFieldValue('startingGold', 0)
					}}
				>
					{_.map(apiObjectToArray(jobClasses), jobClass => (
						<option
							key={jobClass.id}
							value={jobClass.id}
						>
							{jobClass.name}
						</option>
					))}
				</select>
			</div>

			<div className="CreatorFormJobClass__group">
				<h4>Alignment:</h4>
				{jobClass.name === "Paladin" ?
					<div>Lawful Good</div>
					:
					<Field 
						className="select" 
						name="alignment" 
						component="select" 
					>
						{alignments.map((alignment, i) => (
							<option 
								value={alignment} 
								key={`alignment${i}`}
							>
								{alignment}
							</option>
						))}
					</Field>
				}
			</div>		

			<div className="CreatorFormJobClass__group--top">
				<h4>Hit Die:</h4>
				<div>{jobClass.hitDie}</div>
			</div>

			<div className="CreatorFormJobClass__group">
				<h4>Starting Gold:</h4>
				<div className="CreatorForm__input-group">
					<input
						className="number-input"
						type="number"
						value={values.startingGold}
						onChange={(e) => setFieldValue('startingGold', Number(e.target.value))}
					/>
					<div> gp</div>
					<button
						className="button"
						type="button"
						onClick={() => setFieldValue('startingGold', calcStartingGold(jobClass.name))}
					>
						Roll
					</button>
				</div>
			</div>

			<div className="CreatorFormJobClass__group--top">
				<h4>Proficiencies:</h4>
				<div>{jobClass.proficiencies.join(", ")}</div>
			</div>

			<div className="CreatorFormJobClass__group--top">
				<h4>Bonus Languages:</h4>
				<Field 
					className="multi-select" 
					name="bonusLanguages" 
					component="select" 
					multiple 
					onChange={(e) => handleMultiSelect(e, setFieldValue)}
				>
					{_.sortBy(race.bonusLanguages).map(language => (
						<option 
							value={language} 
							key={language}
						>
							{language}
						</option>
					))}
				</Field>                                    
			</div>

			<div className="CreatorFormJobClass__group">
				<h4>Deity:</h4>
				<Field 
					className="text-input" 
					name="deity" 
					placeholder="(Optional) Enter the name of a Deity" 
				/>                                  
			</div>
		
			{jobClass.name === 'Wizard' && (
				<div className="CreatorFormJobClass__group">
					<h4>School:</h4>
					<Field 
						className="select" 
						name="school" 
						component="select" 
						onChange={(e) => {
							if (e.target.value === 'Universal') { 
								setFieldError('prohibitedSchools', undefined)	
								setFieldValue('prohibitedSchools', [])
							}						
							handleChange(e)
						}}
					>
						{schools.map((school, i) => (
							<option 
								value={school} 
								key={`school${i}`}
							>
								{school}
							</option>
						))}
					</Field>
				</div>
			)}

			{jobClass.name === 'Wizard' && school !== 'Universal' && (
				<>
					<div className="CreatorFormJobClass__group--top">
						<h4>Prohibited School(s):</h4>
						<Field 
							className="multi-select"
							name="prohibitedSchools" 
							component="select" 
							multiple 
							validate={validateProhibitedSchools(school)}
							onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
						>
							{_.difference(schools, [school, 'Universal', 'Divination']).map((school, i) => (
								<option 
									value={school} 
									key={`prohibited${i}`}
								>
									{school}
								</option>
							))}
						</Field>
					</div>
					<ErrorMessage className="CreatorForm__error" name="prohibitedSchools" component="div"/>
				</>
			)}
		</div>
	)
}

export default CreatorFormJobClass
