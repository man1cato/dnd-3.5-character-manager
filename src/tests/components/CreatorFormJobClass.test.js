import React from 'react'
import { shallow } from 'enzyme'
import _ from 'lodash'

import CreatorFormJobClass from '../../components/CreatorFormJobClass'
import { apiData } from '../utils/utils'
import { characterOne } from '../utils/seedDatabase'


let wrapper, props

beforeAll(async () => {
   const api = await apiData()
   
   props = {
      values: { 
         race: characterOne.race,
         jobClass: characterOne.jobClass 
      },
      races: api.races,
      jobClasses: api.jobClasses
   }
})

beforeEach(() => {
   wrapper = shallow(<CreatorFormJobClass {...props} />)
})

test('should render CreatorFormJobClass', () => {
	expect(wrapper).toMatchSnapshot()
})

