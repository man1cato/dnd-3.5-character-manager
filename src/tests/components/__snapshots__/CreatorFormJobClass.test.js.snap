// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CreatorFormJobClass 1`] = `
<Fragment>
  <div
    className="form-group--35"
  >
    <h4>
      Job Class:
    </h4>
    <FormikConnect(FieldInner)
      className="select"
      component="select"
      name="jobClass"
      onChange={[Function]}
    >
      <option
        key="jobClass0"
        value="recbxccm8pp3fTmQw"
      >
        Barbarian
      </option>
      <option
        key="jobClass1"
        value="rec0GDUzwsKLYEyqM"
      >
        Bard
      </option>
      <option
        key="jobClass2"
        value="recawEdTMnIzdNnwR"
      >
        Cleric
      </option>
      <option
        key="jobClass3"
        value="recz6owTBNig43RWq"
      >
        Druid
      </option>
      <option
        key="jobClass4"
        value="reckjdfZ5f86z4Rtz"
      >
        Fighter
      </option>
      <option
        key="jobClass5"
        value="recVqqb9HsZRyEAC8"
      >
        Monk
      </option>
      <option
        key="jobClass6"
        value="recjjJTiERQ1ocuNh"
      >
        Paladin
      </option>
      <option
        key="jobClass7"
        value="recKnUPMQEmoCGcag"
      >
        Ranger
      </option>
      <option
        key="jobClass8"
        value="recQrfCKtksnnSsaZ"
      >
        Rogue
      </option>
      <option
        key="jobClass9"
        value="recniRmHZzspPDAFn"
      >
        Sorcerer
      </option>
      <option
        key="jobClass10"
        value="recBCWC99ohXXKqwL"
      >
        Wizard
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <div
    className="form-group--35"
  >
    <h4>
      Alignment:
    </h4>
    <FormikConnect(FieldInner)
      className="select"
      component="select"
      name="alignment"
    >
      <option
        key="alignment0"
        value="Lawful Good"
      >
        Lawful Good
      </option>
      <option
        key="alignment1"
        value="Neutral Good"
      >
        Neutral Good
      </option>
      <option
        key="alignment2"
        value="Chaotic Good"
      >
        Chaotic Good
      </option>
      <option
        key="alignment3"
        value="Lawful Neutral"
      >
        Lawful Neutral
      </option>
      <option
        key="alignment4"
        value="True Neutral"
      >
        True Neutral
      </option>
      <option
        key="alignment5"
        value="Chaotic Neutral"
      >
        Chaotic Neutral
      </option>
      <option
        key="alignment6"
        value="Lawful Evil"
      >
        Lawful Evil
      </option>
      <option
        key="alignment7"
        value="Neutral Evil"
      >
        Neutral Evil
      </option>
      <option
        key="alignment8"
        value="Chaotic Evil"
      >
        Chaotic Evil
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <div
    className="form-group--35 align-top"
  >
    <h4>
      Hit Die:
    </h4>
    <div>
      D4
    </div>
  </div>
  <div
    className="form-group--35 align-top"
  >
    <h4>
      Proficiencies:
    </h4>
    <div>
      Club, Dagger, Light Crossbow, Heavy Crossbow, Quarterstaff
    </div>
  </div>
  <div
    className="form-group--35 align-top"
  >
    <h4>
      Bonus Languages:
    </h4>
    <FormikConnect(FieldInner)
      className="multi-select"
      component="select"
      multiple={true}
      name="bonusLanguages"
      onChange={[Function]}
    >
      <option
        key="Draconic"
        value="Draconic"
      >
        Draconic
      </option>
      <option
        key="Gnoll"
        value="Gnoll"
      >
        Gnoll
      </option>
      <option
        key="Gnome"
        value="Gnome"
      >
        Gnome
      </option>
      <option
        key="Goblin"
        value="Goblin"
      >
        Goblin
      </option>
      <option
        key="Orc"
        value="Orc"
      >
        Orc
      </option>
      <option
        key="Sylvan"
        value="Sylvan"
      >
        Sylvan
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <div
    className="form-group--35"
  >
    <h4>
      Deity:
    </h4>
    <FormikConnect(FieldInner)
      className="text-input"
      name="deity"
      placeholder="(Optional) Enter the name of a Deity"
    />
  </div>
  <div
    className="form-group--35"
  >
    <h4>
      School:
    </h4>
    <FormikConnect(FieldInner)
      className="select"
      component="select"
      name="school"
      onChange={[Function]}
    >
      <option
        key="school0"
        value="Abjuration"
      >
        Abjuration
      </option>
      <option
        key="school1"
        value="Clairsentience"
      >
        Clairsentience
      </option>
      <option
        key="school2"
        value="Conjuration"
      >
        Conjuration
      </option>
      <option
        key="school3"
        value="Divination"
      >
        Divination
      </option>
      <option
        key="school4"
        value="Enchantment"
      >
        Enchantment
      </option>
      <option
        key="school5"
        value="Evocation"
      >
        Evocation
      </option>
      <option
        key="school6"
        value="Illusion"
      >
        Illusion
      </option>
      <option
        key="school7"
        value="Metacreativity"
      >
        Metacreativity
      </option>
      <option
        key="school8"
        value="Necromancy"
      >
        Necromancy
      </option>
      <option
        key="school9"
        value="Psychokinesis"
      >
        Psychokinesis
      </option>
      <option
        key="school10"
        value="Psychometabolism"
      >
        Psychometabolism
      </option>
      <option
        key="school11"
        value="Psychoportation"
      >
        Psychoportation
      </option>
      <option
        key="school12"
        value="Telepathy"
      >
        Telepathy
      </option>
      <option
        key="school13"
        value="Transmutation"
      >
        Transmutation
      </option>
      <option
        key="school14"
        value="Universal"
      >
        Universal
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <div
    className="form-group--35 align-top"
  >
    <h4>
      Prohibited School(s):
    </h4>
    <FormikConnect(FieldInner)
      className="multi-select"
      component="select"
      multiple={true}
      name="prohibitedSchools"
      onChange={[Function]}
      validate={[Function]}
    >
      <option
        key="prohibited0"
        value="Abjuration"
      >
        Abjuration
      </option>
      <option
        key="prohibited1"
        value="Clairsentience"
      >
        Clairsentience
      </option>
      <option
        key="prohibited2"
        value="Conjuration"
      >
        Conjuration
      </option>
      <option
        key="prohibited3"
        value="Enchantment"
      >
        Enchantment
      </option>
      <option
        key="prohibited4"
        value="Evocation"
      >
        Evocation
      </option>
      <option
        key="prohibited5"
        value="Illusion"
      >
        Illusion
      </option>
      <option
        key="prohibited6"
        value="Metacreativity"
      >
        Metacreativity
      </option>
      <option
        key="prohibited7"
        value="Necromancy"
      >
        Necromancy
      </option>
      <option
        key="prohibited8"
        value="Psychokinesis"
      >
        Psychokinesis
      </option>
      <option
        key="prohibited9"
        value="Psychometabolism"
      >
        Psychometabolism
      </option>
      <option
        key="prohibited10"
        value="Psychoportation"
      >
        Psychoportation
      </option>
      <option
        key="prohibited11"
        value="Telepathy"
      >
        Telepathy
      </option>
      <option
        key="prohibited12"
        value="Transmutation"
      >
        Transmutation
      </option>
    </FormikConnect(FieldInner)>
  </div>
  <FormikConnect(ErrorMessageImpl)
    className="form-group--error"
    component="div"
    name="prohibitedSchools"
  />
</Fragment>
`;
